### PR TITLE
Development

### PR DESCRIPTION
**Pull request by SantiNoto**
WIP, not ready for review
Want to migrate the parse part from thift to C, right now just testing thrift.

### CODE REVIEW

File: thrift_tools/thrift_struct.py [hunk 1]
```diff
@@ -69,19 +69,15 @@ def read(cls,
              read_values=False):
         fields = []
         nfields = 0
-
```
Comment (rgs1) [inline thrift_tools/thrift_struct.py]: can you undo all the lines you removed from this file?

File: thrift_tools/thrift_message.py [hunk 2]
```diff
@@ -93,36 +93,39 @@ def read(cls, data,
...
                 proto = protocol(trans)
-
+       
         # unpack the message
         method, mtype, seqid = proto.readMessageBegin()
         mtype = cls.message_type_to_str(mtype)
-
         if len(method) == 0 or method.isspace() or method.startswith(' '):
-            raise ValueError('no method name')
+            raise ValueError('no method nme')
```
Comment (rgs1) [inline thrift_tools/thrift_message.py]: undo this change

File: thrift_tools/thrift_message.py [hunk 3]
```diff
@@ -132,21 +135,19 @@ def read(cls, data,
...
             proto,
             max_fields,
             max_list_size,
             max_map_size,
             max_set_size,
             read_values)
-
         proto.readMessageEnd()
-
+        msglen = trans._buffer.tell()
```
Comment (rgs1) [inline thrift_tools/thrift_message.py]: why was this moved around?

File: thrift_tools/thrift_message.py [hunk 4]
```diff
@@ -93,36 +93,39 @@ def read(cls, data,
...
+                            proto.trans._buffer.tell()
+                            )
+                proto.trans._buffer.seek(args._length)
+               
             except:
                 # reset stream, maybe it's not finagle-thrift
                 trans = TTransport.TMemoryBuffer(data)
                 proto = protocol(trans)
-
+
```
Comment (rgs1) [inline thrift_tools/thrift_message.py]: remove the extra whitespace/tab that was added

File: thrift_tools/tests/test_trivial.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    def setUp(self):
+    	pass
+
+    def tearDown(self):
+        pass
+
+    def test_trivial(self):
+    	testInt=test_C.test_C("12,23,3424")
+    	self.assertEquals(testInt,[12,23,3424])
+    	#testString=thrift_struct_read_module.thrift_struct_read_module("asdfs")
```
Comment (rgs1) [inline thrift_tools/tests/test_trivial.py]: this file is probably not needed anymore

File: thrift_tools/tests/test_basic.py [hunk 6]
```diff
@@ -70,7 +70,7 @@ def test_finagle(self):
         self.assertEquals(msg.method, 'search')
         self.assertEquals(msg.type, 'reply')
         self.assertEquals(msg.args, ThriftStruct([ThriftField('list', 0, ['one', 'two', 'three'])]))
-
+"""
```
Comment (rgs1) [inline thrift_tools/tests/test_basic.py]: why is this commented out?

File: C_functions/thrift_message_read_module.c [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        }
+}
+//just some functions to make C modules callable from python.
+static PyMethodDef thrift_message_read_module_Methods[] = {
+  {"thrift_message_read_module", thrift_message_read_module, METH_VARARGS},
+  {NULL, NULL}
+};
+
+PyMODINIT_FUNC
+initthrift_message_read_module(void)
```
Comment (rgs1) [inline C_functions/thrift_message_read_module.c]: rename this to `init_thrift_message_read_module`

File: C_functions/thrift_message_read_module.c [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        	strcpy(messageType,"exception");
+            return messageType;
+        }
+        else if (mtype == ONEWAY){
+             messageType=malloc(sizeof(char)*(strlen("oneway")+1));
+        	strcpy(messageType,"oneway");
+            return messageType;
+        }
+        else{
+    	PyErr_SetString(PyExc_ValueError,"Unknown message type");
```
Comment (rgs1) [inline C_functions/thrift_message_read_module.c]: indentation seems off?

File: C_functions/thrift_message_read_module.c [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+}
+void readMessageEnd(char *data,char *proto){
+	if(!strcmp(proto,"TJSONProtocol")){
+		readJSONArrayEnd(data);
+	}
+	if(!strcmp(proto,"TCompactProtocol")){
+		if(state!=CLEAR)
+            PyErr_SetString(PyExc_ValueError,"Wrong State");
+        if(*containersStack!=NULL)
+            PyErr_SetString(PyExc_ValueError,"structs stack is not empty");
```
Comment (rgs1) [inline C_functions/thrift_message_read_module.c]: indentation

File: thrift_tools/tests/test_basic.py (hunk 6 above)
Comment (SantiNoto) [inline thrift_tools/tests/test_basic.py]: I dont remember changing that.

File: native/thrift_message_read_module.c [hunk 10]
```diff
@@ -0,0 +1,328 @@
+#include <Python.h>
+#include <string.h>
+#include "thrift_struct_read_module.h"
+#include "thrift_consts.h"
+#include "thrift_utils.h"
+
+char *readMessageBegin(char *,char *,int *,int *);
+void readMessageEnd(char*,char*);
+char *message_type_to_str(int);
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: these should be static (since they are only used in this file, thus the symbols should not be public)

File: native/thrift_message_read_module.c [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+    token=strtok(s, "'");
+    token=strtok(NULL, "'");
+    data=malloc(sizeof(char)*(strlen(token)+1));
+    strcpy(data, token);
+    token=strtok(NULL, ",");
+    proto=token;
+    token=strtok(NULL, ",");
+    finagle_thrift=(strcmp(token, " True"))?0:1;
+    token=strtok(NULL, ")");
+    read_values=(strcmp(token, " True"))?0:1;
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: nit: coding style, space between `=` on all these assignments

File: thrift_tools/thrift_message.py [hunk 12]
```diff
@@ -104,14 +104,14 @@ def read(cls, data,
...
-                header = ThriftStruct.read(
+            try:   
+                header =ThriftStruct.read(
                     proto,
                     max_fields,
                     max_list_size,
                     max_map_size,
                     max_set_size,
-                    read_values)
+                    read_values,)
```
Comment (rgs1) [inline thrift_tools/thrift_message.py]: why?

File: native/thrift_struct_read_module.c [hunk 13]
```diff
@@ -0,0 +1,1851 @@
...
+    }
+    if (type == 10) {
+        return SET;
+    }
+    if (type==11) {
+        return MAP;
+    }
+    if (type == 12) {
+        return STRUCT; 
+    }
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: nit: it'll be more readable to use a switch/case statement here instead of all the ifs

File: native/thrift_struct_read_module.c [hunk 14]
```diff
@@ -0,0 +1,1851 @@
...
+        }
+        readSetEnd(data, proto);
+    }
+    else if (ttype == LIST) {
+        readListBegin(data, proto, &etype, &size);
+        for (i = 0; i < size; i++) {
+            skip(data, proto, ktype);
+        }
+        readListEnd(data, proto);
+    }
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: nit: a switch statement is probably nicer here

File: native/thrift_struct_read_module.c [hunk 15]
```diff
@@ -0,0 +1,1851 @@
...
+        }
+
+        if (state != CONTAINER_READ) {
+            PyErr_SetString(PyExc_ValueError, "Wrong state");
+            return;
+        }
+        temp = popContainers();
+        PyDict_SetItemString(globalVars,"state",Py_BuildValue("i",temp));
+    }
+    return;
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: nit: no need to use `return` here

File: native/thrift_struct_read_module.c [hunk 16]
```diff
@@ -0,0 +1,1851 @@
...
+    }
+}
+
+void
+readListEnd(char *data, char *proto)
+{
+    if (!strcmp(proto, "TJSONProtocol")) {
+        readJSONArrayEnd(data);
+    }
+    if (!strcmp(proto, "TCompactProtocol")) {
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: this should be else if to make it more clear that it's this branch or the previous one (not both)

File: native/thrift_struct_read_module.c [hunk 17]
```diff
@@ -0,0 +1,1851 @@
...
+        }
+
+        if (state != CONTAINER_READ) {
+            PyErr_SetString(PyExc_ValueError, "Wrong state");
+            return ;
+        }
+        temp = popContainers();
+        PyDict_SetItemString(globalVars,"state",Py_BuildValue("i",temp));
+    }
+    return;
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: return is superfluous here

File: native/thrift_struct_read_module.c [hunk 18]
```diff
@@ -0,0 +1,1851 @@
...
+char *readJSONString(char*, int);
+double readJSONDouble(char*);
+void pushContext(int, int, int);
+void popContext();
+
+int readSize(char*);
+int readVarint(char*);
+int fromZigZag(int);
+int getTType(int);
+int readerZigZag(char*);
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: all of these methods are private to this file, so they should be static. See: 

https://stackoverflow.com/questions/558122/what-is-a-static-function

File: thrift_tools/thrift_message.py [hunk 19]
```diff
@@ -104,14 +104,14 @@ def read(cls, data,
         # See: http://git.io/vsziG
         header = None
         if finagle_thrift:
-            try:
-                header = ThriftStruct.read(
+            try:
```
Comment (rgs1) [inline thrift_tools/thrift_message.py]: extra whitespace or tab

File: thrift_tools/thrift_message.py [hunk 20]
```diff
@@ -93,7 +93,7 @@ def read(cls, data,
         # do we have enough data?
         if len(data) < cls.MIN_MESSAGE_SIZE:
             raise ValueError('not enough data')
-
+
```
Comment (rgs1) [inline thrift_tools/thrift_message.py]: extra tab

File: thrift_tools/thrift_struct.py [hunk 21]
```diff
@@ -287,4 +287,4 @@ def as_dict(self):
             'field_id': self.field_id,
             'field_type': self.field_type,
             'value': value
-        }
+        }
```
Comment (rgs1) [inline thrift_tools/thrift_struct.py]: newline was here before

File: native/setup.py [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+module2 = Extension('thrift_struct_read_module', 
+                    sources = ['thrift_struct_read_module.c'])
+
+setup (name = 'thrift_struct_read_module',
+       version = '1.0',
+       description = 'This is the port from the read function in thrift_struct',
+       ext_modules = [module2]
+       )
+module3 = Extension('thrift_message_read_module',
+                    #extra_link_args = ['thrift_utils.c'],
```
Comment (rgs1) [inline native/setup.py]: remove this line if it's not used

File: native/thrift_consts.h [hunk 23]
```diff
@@ -0,0 +1,65 @@
+#define STOP 0
```
Comment (rgs1) [inline native/thrift_consts.h]: please add include guards, i.e.:

```
#ifndef THRIFT_CONSTS_H
... body of the include
#endif
```

see: https://stackoverflow.com/questions/1653958/why-are-ifndef-and-define-used-in-c-header-files

File: native/thrift_message_read_module.c [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+        return NULL;
+    }
+    for (i = 0; i < strlen(method); i++) {
+        if (method[i] < 33 || method[i] > 127) {
+            PyErr_SetString(PyExc_ValueError, "invalid method name");
+            return NULL;
+        }
+    }
+    //Py_BEGIN_ALLOW_THREADS
+    msg=thrift_struct_read(data,
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: nit: coding style, spaces around `=`

File: native/thrift_utils.c [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+    if (val >= 0) {
+        return 0;
+    }
+    int version = val & BINARY_PROTOCOL_VERSION_MASK;
+    return (version == BINARY_PROTOCOL_VERSION_1)?1:0;
+}
+
+int
+is_json_protocol(char *data)
+{
```
Comment (rgs1) [inline native/thrift_utils.c]: protection: you should check that `strlen(data) >= 2` otherwise this could SEGFAULT

File: native/thrift_utils.c [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+    if (data[0] == '\\' && data[1] == 'x') {
+        return !(strcmp(aux, "0x82"));
+    }
+    return (data[0] == strtol("0x82", NULL, 16))?1:0;
+}
+
+int
+is_binary_protocol(char *data)
+{
+    int val;
```
Comment (rgs1) [inline native/thrift_utils.c]: protection: check the min length for data

File: native/thrift_utils.c [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    if (fallbackProtocol == NULL) {
+        return NULL;
+    }
+    return fallbackProtocol;
+}
+
+int
+is_compact_protocol(char *data)
+{
```
Comment (rgs1) [inline native/thrift_utils.c]: check the min length for data

File: native/thrift_message_read_module.c [hunk 28]
```diff
@@ -0,0 +1,328 @@
...
+    strcpy(data, token);
+    token = strtok(NULL, ",");
+    proto = token;
+    token = strtok(NULL, ",");
+    finagle_thrift = (strcmp(token, " True"))?0:1;
+    token = strtok(NULL, ")");
+    read_values = (strcmp(token, " True"))?0:1;
+    
+    if (strcmp(proto, " None") == 0) {
+        protocol = detect_protocol(data, "TBinaryProtocol");
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: when will this the memory allocated for `protocol` be freed?

File: native/thrift_message_read_module.c [hunk 29]
```diff
@@ -0,0 +1,338 @@
...
+                                    MAX_LIST_SIZE,
+                                    MAX_MAP_SIZE,
+                                    MAX_SET_SIZE,
+                                    read_values);
+    }
+
+    method = readMessageBegin(data, protocol, &mtype, &seqid);
+    if (method == NULL || strlen(method) == 0 || method[0] == ' ') {
+        PyErr_SetString(PyExc_ValueError, "no method name");
+        return NULL;
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: if you return here... `protocol` we'll never be freed (memory leak)

File: native/thrift_message_read_module.c [hunk 30]
```diff
@@ -0,0 +1,338 @@
...
+    }
+    if (strlen(method) > MAX_METHOD_LENGTH) {
+        PyErr_SetString(PyExc_ValueError, "method name too long");
+        return NULL;
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: same thing

File: native/thrift_message_read_module.c [hunk 31]
```diff
@@ -0,0 +1,338 @@
...
+    }
+    for (i = 0; i < strlen(method); i++) {
+        if (method[i] < 33 || method[i] > 127) {
+            PyErr_SetString(PyExc_ValueError, "invalid method name");
+            return NULL;
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: same thing

File: native/thrift_message_read_module.c [hunk 32]
```diff
@@ -0,0 +1,338 @@
...
+                            protocol,
+                            MAX_FIELDS,
+                            MAX_LIST_SIZE,
+                            MAX_MAP_SIZE,
+                            MAX_SET_SIZE,
+                            read_values);
+    readMessageEnd(data, protocol);
+    byte = PyDict_GetItemString(globalVars,"byteCounter");
+    if (!PyArg_Parse(byte, "i", &byteCounter)) {
+        return NULL;
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: same thing

File: native/thrift_message_read_module.c [hunk 33]
```diff
@@ -0,0 +1,338 @@
...
+                                                   smtype,
+                                                   seqid,
+                                                   msg,
+                                                   header,
+                                                   msglen));
+    PyTuple_SetItem(result, 1, Py_BuildValue("i", msglen));
+    
+    free(method);
+    free(smtype);
+    free(data);
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: `free(protocol)`

File: native/thrift_message_read_module.c [hunk 34]
```diff
@@ -0,0 +1,338 @@
...
+        }
+        else if (mtype == ONEWAY) {
+            messageType = malloc(sizeof(char)*(strlen("oneway")+1));
+            strcpy(messageType, "oneway");
+            return messageType;
+        }
+        else {
+            return NULL;
+        }
+}
```
Comment (rgs1) [inline native/thrift_message_read_module.c]: when returning constant strings it's better to avoid malloc(), i.e.:

https://gist.github.com/rgs1/8928c8751042274c4707a24f5b5a918e

File: native/thrift_struct_read_module.h [hunk 35]
```diff
@@ -0,0 +1,15 @@
+#ifndef THRIFT_STRUCT_READ_MODULE_H
```
Comment (rgs1) [inline native/thrift_struct_read_module.h]: the second line should be `#define THRIFT_STRUCT_READ_MODULE_H`

see: https://en.wikipedia.org/wiki/Include_guard

File: native/thrift_utils.c [hunk 36]
```diff
@@ -0,0 +1,105 @@
...
+    if (is_json_protocol(data)) {
+        protocol = malloc(sizeof(char)*(strlen("TJSONProtocol")+1));
+        strcpy(protocol, "TJSONProtocol");
+        return protocol;
+    }
+    if (fallbackProtocol == NULL) {
+        return NULL;
+    }
+    return fallbackProtocol;
+}
```
Comment (rgs1) [inline native/thrift_utils.c]: same as `message_type_to_str()`, avoid `malloc()` by using something like:

https://gist.github.com/rgs1/8928c8751042274c4707a24f5b5a918e

File: native/thrift_utils.c [hunk 37]
```diff
@@ -0,0 +1,105 @@
...
+
+int
+is_compact_protocol(char *data)
+{
+    char aux[5];
+    if (strlen(data) < 1) {
+        return -1;
```
Comment (rgs1) [inline native/thrift_utils.c]: you should return 0... otherwise this function will evaluation as true (i.e.: -1 evals as true).

File: native/thrift_utils.c [hunk 38]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+int
+is_binary_protocol(char *data)
+{
+    int val;
+    int version;
+
+    if (strlen(data) < 1) {
+        return -1;
```
Comment (rgs1) [inline native/thrift_utils.c]: this evaluates as true...

File: native/thrift_utils.c [hunk 39]
```diff
@@ -0,0 +1,105 @@
...
+    
+    version = val & BINARY_PROTOCOL_VERSION_MASK;
+    return (version == BINARY_PROTOCOL_VERSION_1)?1:0;
+}
+
+int
+is_json_protocol(char *data)
+{
+    if (strlen(data) < 2) {
+        return -1;
```
Comment (rgs1) [inline native/thrift_utils.c]: this evaluates as true...

File: native/thrift_utils.h [hunk 40]
```diff
@@ -0,0 +1,10 @@
+#ifndef THRIFT_UTILS_H
```
Comment (rgs1) [inline native/thrift_utils.h]: `#define THRIFT_UTILS_H`

File: native/thrift_struct_read_module.c [hunk 41]
```diff
@@ -0,0 +1,1923 @@
...
+    Py_ssize_t structsTop;
+    PyObject *globalVars;
+    struct structs temp;
+        
+    globalVars = PyThreadState_GetDict();
+    
+    structsStack = PyDict_GetItemString(globalVars, "structsStack");
+    py_size = PyDict_GetItemString(globalVars, "structsStackTop");
+    if (!PyArg_Parse(py_size, "n", &structsTop)) {
+            return temp;
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: temp is returned unassigned here...

File: native/thrift_struct_read_module.c [hunk 42]
```diff
@@ -0,0 +1,1923 @@
+#include <Python.h>
+#include <string.h>
+#include <math.h>
+#include "thrift_consts.h"
+
+
+/*some structs to handle stacks needed in TCompactProtocol and TJSONProtocol*/
+
+struct structs
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: maybe a more descriptive name than `structs`? (which is confusing because `struct` is a reserved word...

File: native/thrift_struct_read_module.c [hunk 43]
```diff
@@ -0,0 +1,1923 @@
...
+            return temp;
+    }
+
+    if (containersTop != -1) {
+        PyArg_ParseTuple(PyList_GetItem(ContainersStack, containersTop), "i", &temp);
+        containersTop--;
+        PyDict_SetItemString(globalVars, "containersStackTop", Py_BuildValue("n", containersTop));
+        return temp;
+    }
+    return temp;
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: don't use `temp` as the return value, maybe use `retval` ?

File: native/thrift_struct_read_module.c [hunk 44]
```diff
@@ -0,0 +1,1923 @@
...
+    Py_ssize_t containersTop;
+    PyObject *globalVars;
+    int temp;
+        
+    globalVars = PyThreadState_GetDict();
+    
+    ContainersStack = PyDict_GetItemString(globalVars, "containersStack");
+    py_size = PyDict_GetItemString(globalVars, "containersStackTop");
+    if (!PyArg_Parse(py_size, "n", &containersTop)) {
+            return temp;
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: `temp` is unassigned here... (also name it `retval`)

File: native/thrift_struct_read_module.c [hunk 45]
```diff
@@ -0,0 +1,1923 @@
...
+            readListEnd(data, proto);
+            break;
+        }
+        default :
+            break;
+    }
+}
+
+static char
+*field_type_to_str(int ftype)
```
Comment (rgs1) [inline native/thrift_struct_read_module.c]: `static const char * const field_type_to_str(int ftype)`

**Comment by rgs1** [comment]
Closing since this isn't actively being worked on.
